File: frontend/src/translator/baidu.ts
import { assert } from '@vueuse/shared';
import { Translator } from './base';

export class BaiduQueryProcessor {
  splited_query_list: string[];
  splited_index: number[];
  filtered_splited_query_list: string[];

  constructor(query_list: string[]) {
    this.splited_query_list = [];
    this.splited_index = [];
    for (const query of query_list) {
      const splited = query.split('\n');
      this.splited_query_list = this.splited_query_list.concat(splited);
      this.splited_index.push(splited.length);
    }
    this.filtered_splited_query_list = this.splited_query_list.filter(
      (q) => q.trim() !== ''
    );
  }

  get(): string[] {
    return this.filtered_splited_query_list;
  }

  private *concat(result_list: string[]) {
    assert(result_list.length == this.splited_query_list.length);
    const recovered: string[] = [];
    for (const result of result_list) {
      recovered.push(result);
      if (recovered.length == this.splited_index[0]) {
        this.splited_index.shift();
        yield recovered.join('\n');
        recovered.length = 0;
      }
    }
  }

  recover(result_list: string[]): string[] {
    assert(result_list.length == this.filtered_splited_query_list.length);
    const inserted_result_list: string[] = [];
    for (const query of this.splited_query_list) {
      if (query.trim() !== '') {
        inserted_result_list.push(result_list.shift()!);
      } else {
        inserted_result_list.push(query);
      }
    }
    return Array.from(this.concat(inserted_result_list));
  }
}

export abstract class BaiduBaseTranslator extends Translator {
  private limit_per_request = 2000;

  private *chunk_query(query: string[]): Generator<string | string[]> {
    const chunked: string[] = [];
    let chunked_size = 0;
    for (const line of query) {
      const line_size = line.length;
      if (line_size + chunked_size <= this.limit_per_request) {
        chunked.push(line);
        chunked_size += line_size;
      } else {
        if (chunked.length >= 0) {
          yield chunked;
          chunked.length = 0;
          chunked_size = 0;
        }
        if (line_size <= this.limit_per_request) {
          chunked.push(line);
          chunked_size += line_size;
        } else {
          yield line;
        }
      }
    }
    if (chunked.length >= 0) {
      yield chunked;
    }
  }

  private chunk_string(query: string): string[] {
    const chunks: string[] = [];
    for (let i = 0; i < query.length; i += this.limit_per_request) {
      chunks.push(query.slice(i, i + this.limit_per_request));
    }
    return chunks;
  }

  abstract inner_translate(query: string): Promise<string[]>;

  async translate(query_list: string[]): Promise<string[]> {
    const processor = new BaiduQueryProcessor(query_list);
    const processed_query_list = processor.get();
    let result_list: string[] = [];
    for (const chunked_query of this.chunk_query(processed_query_list)) {
      if (typeof chunked_query === 'string') {
        const chunked_translated = this.chunk_string(chunked_query)
          .map(async (chunked_string) => {
            return await this.inner_translate(chunked_string);
          })
          .flat()
          .join('');
        result_list.push(chunked_translated);
      } else {
        const chunked_translated = await this.inner_translate(
          chunked_query.join('\n')
        );
        result_list = result_list.concat(chunked_translated);
      }
    }

    const recovered_result_list = processor.recover(result_list);
    if (recovered_result_list.length != query_list.length) {
      throw Error('Baidu translator error');
    }
    return recovered_result_list;
  }
}

File: backend/app/translator/baidu.py
from abc import abstractmethod
from typing import Iterator, List

from app.translator.base import Translator


class BaiduQueryProcessor:
    def __init__(self, query_list: List[str]):
        self.splited_query_list: List[str] = []
        self.splited_index: List[int] = []

        for query in query_list:
            splited = query.split("\n")
            self.splited_query_list += splited
            self.splited_index.append(len(splited))

        self.filtered_splited_query_list = [
            q for q in self.splited_query_list if q.strip()
        ]

    def get(self) -> List[str]:
        return self.filtered_splited_query_list

    def _concat(self, result_list: List[str]):
        assert len(result_list) == len(self.splited_query_list)
        recovered = []
        for result in result_list:
            recovered.append(result)
            if len(recovered) == self.splited_index[0]:
                self.splited_index.pop(0)
                yield "\n".join(recovered)
                recovered = []

    def recover(self, result_list: List[str]) -> List[str]:
        assert len(result_list) == len(self.filtered_splited_query_list)
        inserted_result_list = []
        for query in self.splited_query_list:
            if query.strip():
                inserted_result_list.append(result_list.pop(0))
            else:
                inserted_result_list.append(query)
        return list(self._concat(inserted_result_list))


class BaiduBaseTranslate(Translator):
    _limit_per_request = 2000

    def _chunk_query(self, query: List[str]) -> Iterator[str | List[str]]:
        chunked = []
        chunked_size = 0
        for line in query:
            line_size = len(line)
            if line_size + chunked_size <= self._limit_per_request:
                chunked.append(line)
                chunked_size += line_size
            else:
                if chunked:
                    yield chunked
                    chunked = []
                    chunked_size = 0
                if line_size <= self._limit_per_request:
                    chunked.append(line)
                    chunked_size += line_size
                else:
                    yield line
        if chunked:
            yield chunked

    def _chunk_string(self, query: str) -> Iterator[str]:
        return [
            query[i : i + self._limit_per_request]
            for i in range(0, len(query), self._limit_per_request)
        ]

    @abstractmethod
    def _inner_translate(
        self,
        query: str,
    ) -> List[str]:
        pass

    def _translate(
        self,
        query_list: List[str],
    ) -> List[str]:
        processor = BaiduQueryProcessor(query_list=query_list)
        processed_query_list = processor.get()

        result_list = []
        for chunked_query in self._chunk_query(processed_query_list):
            if isinstance(chunked_query, str):
                buffer_result_list = []
                for chunked_string in self._chunk_string(chunked_query):
                    chunked_translated = self._inner_translate(
                        query=chunked_string,
                    )
                    buffer_result_list += chunked_translated
                result_list.append("".join(buffer_result_list))
            else:
                chunked_translated = self._inner_translate(
                    query="\n".join(chunked_query),
                )
                result_list += chunked_translated

        recovered_result_list = processor.recover(result_list=result_list)

        if len(recovered_result_list) != len(query_list):
            raise Exception("Baidu translator error")

        return recovered_result_list
